resolwe/flow/managers/base.py

```diff
@@ -87,11 +180,117 @@ def _include_environment_variables(self, program):
+        output_path = os.path.join(getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR', ''), str(data_id))
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.mkdir(output_path, mode=dir_mode)
+        # os.mkdir is not guaranteed to set the given mode
+        os.chmod(output_path, dir_mode)
+        return output_path
+
+    def _prepare_executor(self, dest_dir, data_id, **kwargs):
+        """Copy executor sources into the destination directory."""
+        settings_dict = dict(kwargs)
+
+        # marshal django settings into a serializable dict
+        for k in dir(settings):
+            if k.startswith('FLOW_'):
+                settings_dict[k] = getattr(settings, k)
+        # add scheduling classes
+        settings_dict['Process_sched_class'] = {
+            k: getattr(Process, k) for k in dir(Process)
+            if k.startswith('SCHEDULING_CLASS_') and isinstance(getattr(Process, k), str)
+        }
+        # add Data status constants
+        settings_dict['Data_status'] = {
+            k: getattr(Data, k) for k in dir(Data)
+            if k.startswith('STATUS_') and isinstance(getattr(Data, k), str)
+        }
+        # required plumbing settings
+        settings_dict['REDIS_CHANNEL_PAIR'] = (
+            MANAGER_EXECUTOR_CHANNELS['result_queue'],
+            MANAGER_EXECUTOR_CHANNELS['result_queue_response'],
+        )
+
+        # extend the settings with whatever the executor wants
+        settings_dict.update(self.executor.make_settings(data_id))
+
+        # save the settings into a json in the data dir
+        json.dump(settings_dict, open(os.path.join(dest_dir, 'settings.json'), 'wt'), cls=SettingsJSONifier)
+
+        # copy executor files to the destination dir
+        import resolwe.flow.executors as rfe
+
+        exec_dir = rfe.__path__[0]
+        dest_dir = os.path.join(dest_dir, 'executors')
+        shutil.copytree(exec_dir, dest_dir)
+        dir_mode = getattr(settings, 'FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.chmod(dest_dir, dir_mode)
+        module_name = type(self.executor).__module__.split('.')
+        module_name = module_name[module_name.index('executors') + 1:]
+        return '.' + '.'.join(module_name) + '.' + type(self.executor).__name__
+
+    def _prepare_script(self, dest_dir, program):
+        """Copy the execution script for this data into the destination directory."""
+        dest_file = os.path.join(dest_dir, 'task.script')
+        open(dest_file, 'wt').write(program)
+        os.chmod(dest_file, stat.S_IRWXU)
+        return 'task.script'
+
+    def override_settings(self, **kwargs):
+        """Set a setting override dictionary in the global manager state for the calling context."""
+        return self._SettingsManager(**kwargs)
+
+    def handle_control_event(self, message):
+        """Handle an event from the channels layer."""
+        cmd = message.content['command']
+
+        if cmd == 'communicate':
```

Comment:
Can you store commands in variables, so you can keep in sync values in executors and listener?

Reply:
Create a separate module (perhaps named `protocols.py`) that is located under `resolwe.flow.managers` that is separately copied to the executor.